### PR TITLE
[ISSUE-464] Scheduler-extender readiness check

### PR DESCRIPTION
**Pull request by safronovD**
### Issue #464 

Implement proposal of implementing scheduler-extender readiness check (based on kube-scheduler restart):
- add readiness probe to extender
- patcher try to recreate kube-scheduler manifest if not changed
- remove kube-scheduler restart check from e2e

## PR checklist
- [x] Add link to the issue
- [x] Choose PR label
- [x] New unit tests added
- [x] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing
_Link to custom CI build_

### CODE REVIEW

**Review by yurkov-anton — APPROVED** [review]

File: docs/proposals/extender-readiness-check.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# Proposal: Scheduler-extender readiness check
+
+Last updated: 08.07.2021
+
+
+## Abstract
+CSI has no way to verify scheduler-extender deployment status.
```
Comment (mishoyama) [inline docs/proposals/extender-readiness-check.md:7]: Abstract sounds confusing. I'd rename to `CSI Operator must inform user about scheduler extender deployment status`

File: docs/proposals/extender-readiness-check.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+The goal - create an opportunity for user to understand: is scheduler-extender installed completely and able to work or not.
+
+## Background
+Currently, there are 2 patching methods depends on platforms.
```
Comment (mishoyama) [inline docs/proposals/extender-readiness-check.md:11]: `Currently CSI Operator doesn't wait for scheduler extender to be completely installed on the system`

File: docs/proposals/extender-readiness-check.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+- For rke/vanilla kubernetes:
+1. Deploy configuration configmap
+2. Deploy csi-baremetal-patcher daemonset.
+In patcher:
+   - Copy configs from cm to local files on host
+   - Update kubernetes-scheduler manifest
+    
+- For Openshift
+1. Deploy specific scheduler configmap
+2. Update Scheduler recourse
```
Comment (mishoyama) [inline docs/proposals/extender-readiness-check.md:22]: Typo - `resource`. Update OpenShift Scheduler Operator configuration.

File: docs/proposals/extender-readiness-check.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+In the both cases patching triggers kube-scheduler restarting process. 
+If kube-scheduler is not restarted, a custom scheduler-extender might not work correctly.
+Restart check is implemented in E2E testing, but it's need to provide this information to user.
+
+## Proposal
+First step - implement CSI CR status field (ready/unready), which is updated in Operator.
+Readiness check - kube-scheduler restarted after patching process.
```
Comment (mishoyama) [inline docs/proposals/extender-readiness-check.md:30]: On all master nodes? How will it work when you add new master node to the cluster?

File: docs/proposals/extender-readiness-check.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+Kube-scheduler restart time - `pod.Status.ContainerStatuses[0].State.Running.StartedAt`
+
+Patching time:
+- For Openshift - Operator checks ConfigMap creation timestamp.
```
Comment (mishoyama) [inline docs/proposals/extender-readiness-check.md:35]: What if CR wasn't updated?

File: docs/proposals/extender-readiness-check.md (hunk 1 above)
Comment (safronovD) [reply docs/proposals/extender-readiness-check.md]: fixed

File: docs/proposals/extender-readiness-check.md (hunk 2 above)
Comment (safronovD) [reply docs/proposals/extender-readiness-check.md]: fixed

File: docs/proposals/extender-readiness-check.md (hunk 3 above)
Comment (safronovD) [reply docs/proposals/extender-readiness-check.md]: fixed

File: docs/proposals/extender-readiness-check.md (hunk 4 above)
Comment (safronovD) [reply docs/proposals/extender-readiness-check.md:67]: add in proposal

File: docs/proposals/extender-readiness-check.md (hunk 5 above)
Comment (safronovD) [reply docs/proposals/extender-readiness-check.md:72]: add in proposal

**Review by mishoyama — CHANGES_REQUESTED** [review]

File: docs/proposals/extender-readiness-check.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+- For Openshift
+1. Deploy specific scheduler configmap
+2. Update OpenShift Scheduler Operator configuration
+
+In the both cases patching triggers kube-scheduler restarting. 
+If kube-scheduler is not restarted, a custom scheduler-extender might not work correctly.
+Restart check is implemented in E2E testing, but it's need to provide this information to user.
+
+## Proposal
+Implement readiness check in scheduler-extender.
```
[the remark this answers is not in the document]
Comment (safronovD) [reply docs/proposals/extender-readiness-check.md]: added

File: docs/proposals/extender-readiness-check.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+![Screenshot](images/extender_flow.png)
+
+CSI Operator behavior for Openshift Platform
```
[the remark this answers is not in the document]
Comment (safronovD) [reply docs/proposals/extender-readiness-check.md:79]: fixed

File: docs/proposals/extender-readiness-check.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+![Screenshot](images/Operator_openshift_flow.png)
+
+Update Timeout ~ 20 sec. Readiness timeout ~ 20 min.
```
[the remark this answers is not in the document]
Comment (safronovD) [reply docs/proposals/extender-readiness-check.md]: it was marked as default

File: docs/proposals/extender-readiness-check.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+In the both cases patching triggers kube-scheduler restarting. 
+If kube-scheduler is not restarted, a custom scheduler-extender might not work correctly.
+Restart check is implemented in E2E testing, but it's need to provide this information to user.
+
+## Proposal
+Implement readiness check in scheduler-extender.
+![Screenshot](images/extender_flow.png)
+
+CSI Operator behavior for Openshift Platform
+![Screenshot](images/Operator_openshift_flow.png)
```
[the remark this answers is not in the document]
Comment (safronovD) [reply docs/proposals/extender-readiness-check.md:80]: fixed

File: docs/proposals/extender-readiness-check.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+Scheduler-extender deployment process:
+
+- For rke/vanilla kubernetes:
+1. Deploy configuration configmap
+2. Deploy csi-baremetal-patcher daemonset.
+In patcher:
+   - Copy configs from cm to local files on host
+   - Update kubernetes-scheduler manifest
+    
+- For Openshift
```
[the remark this answers is not in the document]
Comment (safronovD) [reply docs/proposals/extender-readiness-check.md:42]: I made some test with other custom extender and wrote in Background about it

File: docs/proposals/extender-readiness-check.md [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+1. Deploy CM
+2. Update schedulers.config.openshift.io CR with CM name
+
+In the both cases patching triggers kube-scheduler restart. 
+If kube-scheduler is not restarted, the extender might not work correctly.
+Restart check is implemented in E2E testing, but it's not stable.
+
+#### Patching limitations
+
+If user wants to add another scheduler-extender, he has to manage configs for CSI extender himself.
```
Comment (mishoyama) [inline docs/proposals/extender-readiness-check.md:39]: Please avoid using words like `he`, `she`, `himself`, etc. https://en.wikipedia.org/wiki/Gender-neutral_language

File: docs/proposals/extender-readiness-check.md [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+Manual patching:
+- For Openshift
+1. Set `scheduler.patcher.enable = false`
```
Comment (mishoyama) [inline docs/proposals/extender-readiness-check.md:43]: It's not clear where to set. Please reference to helm chart value. Don't we need use platform parameter instead?

**Review by mishoyama — APPROVED** [review]

File: docs/proposals/extender-readiness-check.md (hunk 11 above)
Comment (safronovD) [reply docs/proposals/extender-readiness-check.md:39]: updated, thank you

File: docs/proposals/extender-readiness-check.md (hunk 12 above)
Comment (safronovD) [reply docs/proposals/extender-readiness-check.md]: updated, remove `scheduler.patcher.enable` in the related PR into Operator

File: cmd/scheduling/extender/main.go [hunk 13]
```diff
@@ -50,6 +54,10 @@ var (
 	metricsAddress = flag.String("metrics-address", "", "The TCP network address where the prometheus metrics endpoint will run"+
 		"(example: :8080 which corresponds to port 8080 on local host). The default is empty string, which means metrics endpoint is disabled.")
 	metricspath = flag.String("metrics-path", "/metrics", "The HTTP path where prometheus metrics will be exposed. Default is /metrics.")
+	healthIP    = flag.String("healthip", base.DefaultHealthIP, "IP for health service")
+	healthPort  = flag.Int("healthport", base.DefaultHealthPort, "Port for health service")
+	nodeName    = flag.String("nodeName", "", "Name of the node with assigned extender pod")
```
Comment (mishoyama) [inline cmd/scheduling/extender/main.go:59]: node name can be obtained from ENV var

File: cmd/scheduling/extender/main.go [hunk 14]
```diff
@@ -50,6 +54,10 @@ var (
...
+	statusFile  = flag.String("statusFile", "", "Path to file from ConfigMap with readiness status")
```
Comment (mishoyama) [inline cmd/scheduling/extender/main.go:60]: Do you expect file path to be configurable?

File: docs/proposals/extender-readiness-check.md (hunk 11 above)
Comment (mishoyama) [reply docs/proposals/extender-readiness-check.md:39]: you still have `he` in patching limitations

**Review by mishoyama — CHANGES_REQUESTED** [review]

File: Makefile.docker [hunk 15]
```diff
@@ -48,8 +48,9 @@ image-controller: base-image-controller
 	cp ./pkg/${CONTROLLER}/Dockerfile ./build/${CONTROLLER}/
 	docker build --network host --force-rm --tag ${REGISTRY}/${PROJECT}-${CONTROLLER}:${TAG} ./build/${CONTROLLER}
 
-image-extender:
+image-extender: download-grpc-health-probe
```
Comment (mishoyama) [inline Makefile.docker:51]: I won't be able to build my project without access to the internet with this dependency.

File: cmd/scheduling/extender/main.go (hunk 13 above)
Comment (safronovD) [reply cmd/scheduling/extender/main.go]: used ENV instead

File: cmd/scheduling/extender/main.go (hunk 14 above)
Comment (safronovD) [reply cmd/scheduling/extender/main.go]: File path is stored in const. Added `isPathingEnable` bool arg

File: Makefile.docker (hunk 15 above)
Comment (safronovD) [reply Makefile.docker]: updated

**Review by mishoyama — APPROVED** [review]

**Pull request merged**